src/Juvix/FrontendContextualise/ModuleOpen/Environment.hs

```diff
@@ -100,3 +159,199 @@ lookupModMap s =
 removeModMap ::
   HasState "modMap" ModuleMap m => Symbol -> m ()
 removeModMap s = Juvix.Library.modify @"modMap" (Map.delete s)
+
+--------------------------------------------------------------------------------
+-- fully resolve module opens
+--------------------------------------------------------------------------------
+
+-- Precondition ∷ old and new context must be in the same context
+-- also terms must be in either the old map or the new map
+-- any place where this is inconsistent will break resolution
+
+-- | @populateModMap@ populates the modMap with all the global opens
+-- in the module
+populateModMap ::
+  WorkingMaps m => m ()
+populateModMap = do
+  old <- get @"old"
+  new <- get @"new"
+  open <- Juvix.Library.ask @"openMap"
+  modM <- get @"modMap"
+  let curr = pure Context.topLevelName <> Context.currentName old
+  case open Map.!? curr of
+    Nothing ->
+      pure ()
+    Just opens ->
+      -- TODO ∷
+      -- explicts and implicits for now work the same
+      put @"modMap" (foldr f modM opens)
+      where
+        f y modMap = foldr addNewSymbol modMap openList
+          where
+            nameSpace =
+              case y of
+                Implicit x -> x
+                Explicit x -> x
+            unpopulate = (() <$)
+            --
+            addNewSymbol x modMap =
+              case unpopulate (Context.lookup (NameSymbol.fromSymbol x) old)
+                <|> unpopulate (Context.lookup (NameSymbol.fromSymbol x) new) of
+                Just () ->
+                  -- if the symbol is in the map, then
+                  -- we don't shadow it....
+                  -- TODO ∷ later throw an error for
+                  --        explicit opens that do this
+                  modMap
+                Nothing ->
+                  Map.insert x nameSpace modMap
+            --
+            openList =
+              -- haskell gives type error if grabList is not typed
+              let grabList ::
```

Comment:
Should this just be a top-level function?

Reply:
I think the limited scope is fine, if we have to use this logic again (we likely won't), then it's an easy move

Reply:
alright